### ISSUE TITLE
Add bcs to move-native

### ISSUE THREAD

**Issue by brson** [solana-labs/move#128]
[bcs](https://docs.rs/bcs/latest/bcs/) is the serialization format used by move, and we'll probably want to continue using it in the port.

It is needed in move-native to implement the native std::bcs::to_bytes function, and will probably be needed by the compiler for storage-related runtime calls.

The to_bytes function is already implemented but commented out because move-native is (currently) a no-std crate, and the bcs crate requires the standard library.

The reason move-native is no-std is twofold:

1) while Solana programs can link to the standard library, some platforms that might want to use move-llvm can not, and I still hope this project might be useful beyond solana. I understand this is potentially a weak argument.
2) last time i tried to link move-native with the standard library I encountered linker errors I couldn't resolve.

Note that the the solana build of move-native will probably link to the standard library even if the rest of move-native does not, as interacting with solana will probably eventually require linking to the solana_program crate, which links to std.

There are a few paths forward:

- look into adding a "std" default feature to the bcs crate, and making it a default feature, able to be turned off. Serialization crates are usually amenable to compiling in a no-std mode. Try to upstream it. This is my ideal, but maybe not practical in the short term.

- We could also keep move-native itself `#![no_std]`, disallowing it to use the standard library, while linking to bcs and letting bcs use std; with the ambition that this discrepancy can be resolved later. Need to solve any remaining linking issues with the rust standard library. I'm ok with this.

- Give up the idea that move-native is no-std. I like this less, but also it might be inevitable that this port is solana-specific.

**Comment by nvjle** [comment]
Now that the compiler proper has reached a near feature-complete state, this issue has risen to the top of the priority queue.

Other projects have already run into the same issue we have-- `bcs` requiring the standard library. At least two forks have recently added `no_std` support:
1. https://github.com/Entropy-Foundation/bcs
2. https://github.com/KeystoneHQ/bcs

Fork 2 appears to be the cleaner implementation as it uses the `core2` crate instead of re-implementing `std` functionality.

If we are amenable to using a different fork than the `zefchain` fork that the crates.io `bcs` points to, then it seems we have a solution for the `bcs` problem. This seems fine to me-- the `zefchain` fork seems dead-- or at least virtually inactive.

In any case, we are to the point where we need an ordinarily built `move-stdlib`, and this is the last blocker.

**Comment by brson** [comment]
Yeah I can hack on this.

**Comment by brson** [comment]
They KeystoneHQ version is nightly only - it turns on the error_in_core feature.

**Comment by brson** [comment]
That might be workable since I think solana's bpf toolchain is nightly and/or sneakily uses nightly features on stable, though we also compile move-native for x86 using a non-nightly compiler, so it would require some conditional compilation to distinguish the two cases.

**Comment by brson** [comment]
I wonder if it is important that the bcs::to_bytes function produces the same results as the real move compiler, or if we can just produce something usable. Within move-native the rust types are not the same as within the move vm, so presumably by default we're going to get some different serializations.

**Comment by brson** [comment]
I've started on this with the Entropy-Foundation fork since it doesn't require nightly. That fork is a bit hacked up though, with a test suite that doesn't compile. I have submitted a build-fix that we need to their fork: https://github.com/Entropy-Foundation/bcs/pull/1

**Comment by nvjle** [comment]
> I've started on this with the Entropy-Foundation fork since it doesn't require nightly. That fork is a bit hacked up though, with a test suite that doesn't compile. I have submitted a build-fix that we need to their fork: [Entropy-Foundation/bcs#1](https://github.com/Entropy-Foundation/bcs/pull/1)

I'm okay with with either fork, although I did like the Keystone version better (cleaner, simpler, less hacky), but I understand not wanting to use nightly. Either one unblocks us and saves us the effort of doing redundant work that others have done.

**Comment by nvjle** [comment]
> I wonder if it is important that the bcs::to_bytes function produces the same results as the real move compiler, or if we can just produce something usable. Within move-native the rust types are not the same as within the move vm, so presumably by default we're going to get some different serializations.

This question occurred to me as well[*]. I believe we'll just need to produce something usable. The main places I see it used are in the data cache (of course) and in the transaction interface. We won't need to interoperate with any other store than our own, so that's no problem. Our transaction interface (calling Move entry points, receiving return values), I think, will already be fully custom and set up serializations the way our runtime expects, so that deserializations will produce our expected values. It may not be as straightforward as things are in the Move VM, but I think it'll work. 
  

[*] I had considered the possibility of just making bcs ser/des wrappers over Borsh. The latter is already used by Solana, very well supported, active, many languages, no_std support, etc. But then I found those bcs forks. On the other hand, if we indeed only need to produce usable serializations, using Borsh under the hood might be a good way to go.

**Comment by brson** [comment]
I have found another issue with the Entropy-Foundation fork - it doesn't support deserialization in no_std mode, and we'll need it.

I'm liking the idea of just using borsh, at least for the moment, and assuming we don't need the serialization to be interoperable outside solana.

**Comment by brson** [comment]
Since borsh doesn't use serde, using it will involve some rewriting of the existing serialization implemantation. Though on the positivie side the runtime wouldn't have to import serde at all, just borsh.

**Comment by brson** [comment]
At the moment I am experimenting with https://docs.rs/postcard/latest/postcard/ just to get something both serde-compatible and no-std compatible running in the test suite. Swapping it out with either bcs or borsh should be straightforward, with borsh requiring rewriting the serialize/deserialize implementations.

I will do so before submitting a pr.

**Comment by nvjle** [comment]
> At the moment I am experimenting with https://docs.rs/postcard/latest/postcard/ just to get something both serde-compatible and no-std compatible running in the test suite. Swapping it out with either bcs or borsh should be straightforward, with borsh requiring rewriting the serialize/deserialize implementations.

SGTM (and +1 from me if we go Borsh).

**Comment by brson** [comment]
I've made some progress on this. Throwing out the old serde impls, which I don't think were structured quite right for our circumstances where we are dealing with a bunch of transmuted reference types. Currently debugging a mysterious abort when serializing vectors.

Using borsh now.

**Comment by nvjle** [comment]
> I've made some progress on this. Throwing out the old serde impls, which I don't think were structured quite right for our circumstances where we are dealing with a bunch of transmuted reference types. Currently debugging a mysterious abort when serializing vectors.
> 
> Using borsh now.

I vaguely remember a weird crash/abort when I was working on the debug formatters for structs in move-native. I've context switched most of the details out of memory, but you might take a look at https://github.com/solana-labs/move/issues/191 and my related patches and see if it might be related. I do remember eye-balling the serializers (as they existed then) thinking that they might suffer from something similar. But by now the serializers have been rewritten in your current work, so it might be of no help. 

I did add a TRACE option to the rbpf-tests runner to help debug the above. It was a tremendous help.

But I'm otherwise glad to hear the progress and Borsh being used.

**Comment by nvjle** [comment]
I located your repository and branch in an effort to take a quick look. Turning on `TRACE`, I see that we crap out either in `vec_destroy` (with the `bcs::test_from_bytes` line commented out in `stdlib-bcs.move`) or `MoveUntypedVector::drop` when it is left in (i.e., as the test would ordinarily be run). We're not dropping a vector somewhere, but I have not dug any further than that. I'll dig further if this doesn't help you right away. 

I also just added this locally to make panics a little more obvious and useful, compared to tracing SBF instructions:
```@@ -2130,6 +2145,7 @@ pub(crate) mod target_defs {
 
     mod globals {
         use alloc::alloc::{GlobalAlloc, Layout};
+        use alloc::format;
         use core::mem::size_of;
         use core::ptr::null_mut;
 
@@ -2137,6 +2153,7 @@ pub(crate) mod target_defs {
 
         #[panic_handler]
         fn panic(info: &core::panic::PanicInfo) -> ! {
+            super::print_string(&format!("{}", info));
             super::abort(PANIC_ABORT_CODE);
         }
```

Now running the test gives a very clear indication:
```thread 'run_test::stdlib-bcs.move' panicked at 'test aborted unexpectedly', language/tools/move-mv-llvm-compiler/tests/rbpf-tests.rs:507:21
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
event 0: "panicked at 'forgot to destroy MoveUntypedVector', src/lib.rs:526:13"
event 1: "0x65, 0x65, 0x65, 0x65, 0x65"
test run_test::stdlib-bcs.move ... FAILED
```

**Comment by nvjle** [comment]
Also, FWIW, disabling the drop panic does allow the test to pass. So I think the actual ser/des is close to working for vector. :+1: 

Question: I was just perusing the Postcard documentation and noticed that it apparently has its own "wire" format-- that is, a binary format. Does this mean that we could just use Postcard exclusively instead of Borsh or any other secondary layer?

**Comment by brson** [comment]
Thanks for debugging.

To your question, I think Postcard is designed for this type of scenario and we could use it, though I have not heard of it until now. I'm not sure what you mean by "secondary layer". Postcard requires serde, while Borsh is self-contained.

**Comment by nvjle** [comment]
> Thanks for debugging.
> 
> To your question, I think Postcard is designed for this type of scenario and we could use it, though I have not heard of it until now. I'm not sure what you mean by "secondary layer". Postcard requires serde, while Borsh is self-contained.

By "second layer" (poor name for lack of something better), I meant an API outside of Postcard (Borsh) which reads/writes its own binary format. It seems that if Postcard has its own serialized binary representation, then Borsh (or any other such thing) would be superfluous. By "binary representation", I'm referring to the serialized data that sits in the "persistent store".

On the other hand, with the standard Solana apps using Borsh, maybe it is necessary. I still don't have a concrete grasp of interoperability of Move and Rust Solana apps, and this may be a prerequisite.  That, and it's almost working.

**Comment by brson** [comment]
The main thing remaining to implement is `vector<struct>`. I will do this next.

**Comment by brson** [comment]
It may help to finish https://github.com/solana-labs/move/issues/220 first since at present we don't know whether some of the more complex serialization routines actually round-trip correctly for the lack of an eq operation.

**Comment by nvjle** [comment]
> It may help to finish #220 first since at present we don't know whether some of the more complex serialization routines actually round-trip correctly for the lack of an eq operation.

I agree-- see comment in #220.
